### PR TITLE
Improvements to std.uni.sicmp and std.uni.icmp

### PR DESCRIPTION
**Pull request by JackStouffer**
The main points:

1. The constraints for `sicmp` were completely wrong because `decode` cannot be called with forward ranges. In the future this could be made to work with forward ranges but for now this needs to be fixed
2. Emphasized the fact that `icmp` can be `@nogc @safe nothrow pure` if used with `byUTF`.
3. Added more detail to `icmp`'s docs

### CODE REVIEW

**Comment by JackStouffer** [comment]
@wilzbach @Hackerpilot what's with the random travis error?

File: std/uni.d [hunk 1]
```diff
@@ -7094,11 +7103,34 @@ int icmp(S1, S2)(S1 str1, S2 str2)
...
+{
     assert(icmp("Rußland", "Russland") == 0);
     assert(icmp("ᾩ -> \u1F70\u03B9", "\u1F61\u03B9 -> ᾲ") == 0);
 }
 
+/**
+ * By using $(XREF utf, byUTF) and its aliases, GC allocations via auto-decoding
+ * and thrown exceptions can be avoided.
+ */
+@safe nothrow pure unittest
```
Comment (DmitryOlshansky) [inline std/uni.d]: add nogc?

**Comment by wilzbach** [comment]
> @wilzbach @Hackerpilot what's with the random travis error?

Should be fixed once #4255 kicks in. Sorry about that :/

File: std/uni.d (hunk 1 above)
Comment (JackStouffer) [inline std/uni.d]: Can't with string literals.

Comment (DmitryOlshansky) [inline std/uni.d]: Ehm.. They are static what's wrong with it?

Comment (JackStouffer) [inline std/uni.d]: Huh, I guess you can add it. I thought that the rules about array literals applied to string literals.

**Comment by JackStouffer** [comment]
@DmitryOlshansky fixed

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

File: std/uni.d [hunk 2]
```diff
@@ -7094,11 +7103,24 @@ int icmp(S1, S2)(S1 str1, S2 str2)
...
 
+/**
+ * By using $(XREF utf, byUTF) and its aliases, GC allocations via auto-decoding
+ * and thrown exceptions can be avoided, making `icmp` `@safe @nogc nothrow pure`.
+ */
+@safe @nogc nothrow pure unittest
+{
+    import std.utf : byDchar;
+
+    assert(icmp("Rußland".byDchar, "Russland".byDchar) == 0);
```
Comment (wilzbach) [inline std/uni.d]: Wait what??? 
In German (my mother tongue) this is not guaranteed to be correct. From Wikipedia:

Thus it helps to distinguish words like Buße (long vowel) 'penance, fine' and Busse (short vowel) 'buses'. 

https://en.m.wikipedia.org/wiki/%C3%9F

Comment (DmitryOlshansky) [inline std/uni.d]: This is caseless comparison and follows rules of Unicode

Comment (wilzbach) [inline std/uni.d]: > This is caseless comparison and follows rules of Unicode

Oh sorry. Overlooked from my phone :/

**Pull request merged**